### PR TITLE
Light Validator

### PR DESCRIPTION
**Pull request by hmalik88**
Validator contract for the `Light` protocol. A typescript class was also added as a pass through for use on the client side. The contract checks for the following:

- Invalid & unauthorized signatures
- Low approvals/balances
- Expired orders
- Used nonces

The contract must be deployed with a `Light` address. A function to allow the owner of the contract to set the `Light` address was also added.

### CODE REVIEW

**Review by dmosites — CHANGES_REQUESTED** [review]
Looking good. Left some small notes and suggest `deployMockContract` for tests.

```
const { waffle } = require('hardhat')
const { deployMockContract } = waffle
```

File: source/light-validator/contracts/LightValidator.sol [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    details.s = s;
+    details.senderWallet = senderWallet;
+    bytes32 hashed = _getOrderHash(details);
+    address signatory = _getSignatory(hashed, v, r, s);
+    uint256 swapFee = details.signerAmount.mul(Light(light).signerFee()).div(
+      Light(light).FEE_DIVISOR()
+    );
+    // Ensure the signatory is not null
+    if (signatory == address(0)) {
+      errors[errCount] = "SIGNATURE_INVALID";
```
Comment (dmosites) [inline source/light-validator/contracts/LightValidator.sol:113]: `INVALID_SIG` (From Light.sol)

Comment (dmosites) [inline source/light-validator/contracts/LightValidator.sol:113]: `INVALID_SIG`

File: source/light-validator/contracts/LightValidator.sol [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      errCount++;
+    }
+    //expiry check
+    if (details.expiry < block.timestamp) {
+      errors[errCount] = "ORDER_EXPIRED";
```
Comment (dmosites) [inline source/light-validator/contracts/LightValidator.sol:118]: `EXPIRY_PASSED`

File: source/light-validator/contracts/LightValidator.sol [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+      errors[errCount] = "SENDER_BALANCE_LOW";
+      errCount++;
+    }
+    if (signerBalance < details.signerAmount + swapFee) {
+      errors[errCount] = "SIGNER_BALANCE_LOW";
+      errCount++;
+    }
+    //nonce check
+    if (Light(light).nonceUsed(details.signerWallet, details.nonce)) {
+      errors[errCount] = "ORDER_TAKEN_OR_CANCELLED";
```
Comment (dmosites) [inline source/light-validator/contracts/LightValidator.sol:162]: `NONCE_ALREADY_USED`

File: source/light-validator/contracts/LightValidator.sol [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    }
+    //expiry check
+    if (details.expiry < block.timestamp) {
+      errors[errCount] = "ORDER_EXPIRED";
+      errCount++;
+    }
+    //if signatory is not the signerWallet, then it must have been authorized
+    if (details.signerWallet != signatory) {
+      if (Light(light).authorized(details.signerWallet) != signatory) {
+        errors[errCount] = "SIGNATURE_UNAUTHORIZED";
```
Comment (dmosites) [inline source/light-validator/contracts/LightValidator.sol:124]: `UNAUTHORIZED`

File: tools/protocols/src/LightValidator.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import { ethers } from 'ethers'
+import { BigNumber } from 'ethers'
+import { chainIds, chainNames } from '@airswap/constants'
+import { LightOrder } from '@airswap/types'
+import * as LightValidatorContract from '@airswap/light-validator/build/contracts/LightValidator.sol/LightValidator.json'
+import * as LightValidatorDeploys from '@airswap/light/deploys.js'
```
Comment (dmosites) [inline tools/protocols/src/LightValidator.ts:6]: `@airswap/light-validator/deploys.js`

**Review by dmosites — CHANGES_REQUESTED** [review]
Would like revert reasons to be the same as those on `Light.sol` and small catch on the TypeScript library. With those in let's just name it `@airswap/validator` and it's ready to go!

File: tools/protocols/src/LightValidator.ts (hunk 5 above)
Comment (hmalik88) [reply tools/protocols/src/LightValidator.ts:6]: with the rename to @airswap/validator, this should then be `@airswap/validator/deploys.js` no?

Comment (dmosites) [reply tools/protocols/src/LightValidator.ts:6]: Correct! Just wanted to be consistent with the current state.

File: source/light-validator/package.json [hunk 6]
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "@airswap/validator",
+  "version": "0.0.0",
```
Comment (dmosites) [inline source/light-validator/package.json:3]: To properly publish over the existing package please set this to `1.2.8`

**Review by dmosites — APPROVED** [review]

**Pull request merged**